### DISCUSSION TITLE
Policy Attachment: Priority/order of generated config when multiple Policies applied to a resource

### DISCUSSION THREAD

**Discussion by sunjayBhatia** [kubernetes-sigs/gateway-api#1498] [General]
If multiple Policy resources target a resource or resource hierarchy, the guidelines on Policy Attachment do not currently have any guidance (or at least a note for implementers) on how to prioritize the generated configuration.

For example, an implementation may provide a Policy that provides rate-limiting functionality and another that provides some authentication mechanism. They may both target an HTTPRoute or Gateway resource.

The order of the generated configuration for the data-plane in this example could have significant impact on resulting behavior for users (e.g. some expecting rate limits to be applied before authentication of a request, or vise versa). In the case of an Envoy data-plane based implementation, this may result in differing orders of [HTTP filters](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/http_filters).

This definitely seems primarily like an implementation-specific issue, at the very least each implementation should be clear about the precedence of the Policies it supports.

This discussion topic is here to ask a few further questions:
- Is there anything we can do in the Policy Attachment spec/guidelines to make this clearer for implementers? Maybe at least stating in the docs that we expect implementers to be clear about Policy order as stated above?
- How do users know what the precedence of Policies they have applied to their resources is? Is it as simple as relying on implementation specific documentation?
- Is there anything we can do in the Policy Attachment spec/guidelines to make this clearer for users?

This sort of falls in a general bucket of issues with [this discussion](https://github.com/kubernetes-sigs/gateway-api/discussions/1489) about how we make Policy resources and their effects observable to users

Other notes:
- HTTPRoute Filters do have an order, which can help with implementing configuration priority, but these don't operate at the scope that immediately makes them applicable to things that might translate to Envoy HTTP filter configuration

**Comment by shaneutt** [comment]
> Is there anything we can do in the Policy Attachment spec/guidelines to make this clearer for implementers? Maybe at least stating in the docs that we expect implementers to be clear about Policy order as stated above?

I think it's fair to indicate in the docs that we expect implementations to document the priority behavior of policies according to their internals.

> How do users know what the precedence of Policies they have applied to their resources is? Is it as simple as relying on implementation specific documentation?

For right now, in the absence of some clear user issues cropping up driving change, I would argue that this is very implementation specific and the documentation thereof should let the users know what to expect.

> Is there anything we can do in the Policy Attachment spec/guidelines to make this clearer for users?

I'm not immediately sure what we could or even should do. This is an area so deep in implementation specific nuances that I don't expect we can provide excellent clarity that's ubiquitous across implementations. The only thing that comes to mind right now is related to this post: https://github.com/kubernetes-sigs/gateway-api/discussions/1489#discussioncomment-4029926. If we had status information in routes which indicated the policies which are under effect, we could potentially add ordering or priority information where the implementations can indicate the priority they will give the policies in the status. It's a big "if" but perhaps something to explore.

---

Separately, but relatedly: if you suggest that users should be able to **influence** priority (rather than the API simply showing the state of priority) it may be reasonable to document an optional concept of user-defined priority (though you still need to document at an implementation level how tie breakers are going to work).

**Reply by sunjayBhatia** [reply]
> I think it's fair to indicate in the docs that we expect implementations to document the priority behavior of policies according to their internals.

yeah, realistically I think this might be the only action item to come out of this discussion because it is very much an implementation issue, but maybe someone has some other useful thoughts here

**Reply by youngnick** [reply]
I think that we should provide recommendations (SHOULD) in the Policy Attachment docs about what status to indicate, and maybe a place/socket/field in Route status for it to live, with the caveats that you need to be very careful about how much you update it.

I'll add something about this to the Policy attachment docs rework I'm working on.
